runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillTimerInternals.java

```diff
@@ -285,13 +308,27 @@ public static TimerData windmillTimerToTimerData(
    * <p>This is necessary because Windmill will deduplicate based only on this tag.
    */
   public static ByteString timerTag(WindmillNamespacePrefix prefix, TimerData timerData) {
-    String tagString =
-        new StringBuilder()
-            .append(prefix.byteString().toStringUtf8()) // this never ends with a slash
-            .append(timerData.getNamespace().stringKey()) // this must begin and end with a slash
-            .append('+')
-            .append(timerData.getTimerId()) // this is arbitrary; currently unescaped
-            .toString();
+    String tagString;
+    // Timers without timerFamily would have timerFamily would be an empty string
+    if ("".equals(timerData.getTimerFamilyId())) {
+      tagString =
+          new StringBuilder()
+              .append(prefix.byteString().toStringUtf8()) // this never ends with a slash
+              .append(timerData.getNamespace().stringKey()) // this must begin and end with a slash
+              .append('+')
+              .append(timerData.getTimerId()) // this is arbitrary; currently unescaped
+              .toString();
+    } else {
+      tagString =
+          new StringBuilder()
+              .append(prefix.byteString().toStringUtf8()) // this never ends with a slash
+              .append(timerData.getNamespace().stringKey()) // this must begin and end with a slash
+              .append('+')
+              .append(timerData.getTimerId()) // this is arbitrary; currently unescaped
+              .append('+')
+              .append(timerData.getTimerFamilyId())
+              .toString();
+    }
```

Comment:
I just realized that timerHoldTag was not updated, which will cause this feature to be incompatible with the previous outputTimestamp feature. This logic needs to be executed there as well (also backwards compatible)

Reply:
> Overall this looks good now! Good job debugging the overlapping timers.
> 
> I think the only thing that remaining will be to add a UsesTimerMap test category so that we can exclude the runners that don't implement this. I'll trigger ValidatesRunner for those runners, though I expect them to break.

Thank you @reuvenlax  for sparing your time to guide me whenever I get stuck. I am adding a new category for TImerMap. Seems like I only need to exclude Flink runner for this category.

Reply:
> actually maybe we don't need a new tag after all, as these tests are surprisingly passing

@reuvenlax , timerHoldTag updated